torch/distributed/launch.py

```diff
@@ -0,0 +1,179 @@
+r"""
+torch.distributed.launch is a helper module that spawns up multiple distributed
+training processes on each of the training nodes.
+
+The utility can be used for single-node distributed training, in which one or
+more processes per node will be spawned. The utility can be used for either
+CPU training or GPU training. If the utility is used for GPU training,
+each distributed process will be operating on a single GPU. This can achieve
+well-improved single-node training performance. It can also be used in
+multi-node distributed training, by spawning up multiple processes on each node
+for well-improved multi-node distributed training performance as well.
+This will especially be benefitial for systems with multiple Infiniband
+interfaces that have direct-GPU support, since all of them can be utilized for
+aggregated communication bandwidth.
+
+In both cases of single-node distributed training or multi-node distributed
+training, this utility will launch the given number of processes per node
+(``--nproc_per_node``). If used for GPU training, this number needs to be less
+or euqal to the number of GPUs on the current system (``nproc_per_node``),
+and each process will be operating on a single GPU from *GPU 0 to
+GPU (nproc_per_node - 1)*.
+
+**How to use this module:**
+
+1. Single-Node multi-process distributed training
+
+  ``python -m torch.distributed.launch --nproc_per_node=NUM_GPUS_YOU_HAVE
+  YOUR_TRAINING_SCRIPT.py (--arg1 --arg2 --arg3 and all other arguments of
+  your training script)``
+
+2. Multi-Node multi-process distributed training: (e.g. two nodes)
+
+    **Node 1**: *(IP: 192.168.1.1, and has a free port: 1234)*
+        ``python -m torch.distributed.launch --nproc_per_node=NUM_GPUS_YOU_HAVE
+        --nnodes=2 --node_rank=0 --master_addr="192.168.1.1"
+        --master_port=1234 YOUR_TRAINING_SCRIPT.py (--arg1 --arg2 --arg3 and
+        all other arguments of your training script)``
+    **Node 2**:
+        ``python -m torch.distributed.launch --nproc_per_node=NUM_GPUS_YOU_HAVE
+        --nnodes=2 --node_rank=1 --master_addr="192.168.1.1"
+        --master_port=1234 YOUR_TRAINING_SCRIPT.py (--arg1 --arg2 --arg3 and
+        all other arguments of your training script)``
+
+3. To look up what optional arguments this module offers:
+
+        >>> python -m torch.distributed.launch --help
+
+**Important Notices:**
+
+1. This utilty and multi-process distributed (single-node or
+multi-node) GPU training currently only achieves the best performance using
+the NCCL distributed backend. Thus NCCL backend is the recommended backend to
+use for GPU training.
```

Comment:
I noticed that we can allocate a rank per node while DistributedDataParallel supports to use all GPU at the node immediately, is use NUM_GPUS_YOU_HAVE ranks per node better than the former?